### ISSUE TITLE
'readany()' returns chunks forever  

### ISSUE THREAD

**Issue by AIGeneratedUsername** [JWCook/aiohttp-client-cache#89]
### The problem

Consider the following code which works fine without **aiohttp-client-cache**:
```python
        async def bytes_parser(response: ClientResponse) -> bytes:
            response_bytes = b""
            while chunk := await response.content.readany():  # Read next data portion for the stream. Returns immediately if internal buffer has a data.
                self.logger.debug(f"A chunk ({chunk}) was read. URL: {url}")
                response_bytes += chunk
```
When I enable **aiohttp-client-cache**, then 'readany()' returns chunks forever:
```
DEBUG    aiohttp_client_cache.backends.base:base.py:86 Pre-cache checks for response from https://httpbin.org/stream-bytes/10: {'allowed method': True, 'allowed status': True, 'not disabled': True, 'not expired': True, 'not filtered': True}
DEBUG    aiohttp_client_cache.backends.base:base.py:137 Cached response found for key: 411ccb04060275cf3bc5a3de7820e4b8b627a0e95ef51c9da8178a915c855f1b
DEBUG    api.py:497 Got response 200 for URL: https://httpbin.org/stream-bytes/10, method: GET, params: None, headers: {'Accept': 'application/octet-stream', 'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/81.0.4044.138 Safari/537.36', 'Referer': 'https://www.google.com/', 'Accept-Language': 'en-US,en;q=0.5', 'Connection': 'keep-alive'}, kwargs: {}
DEBUG    api.py:603 A chunk (b' \xc1o\xdfu9?\x02C\xa5') was read. URL: https://httpbin.org/stream-bytes/10
DEBUG    api.py:603 A chunk (b' \xc1o\xdfu9?\x02C\xa5') was read. URL: https://httpbin.org/stream-bytes/10
DEBUG    api.py:603 A chunk (b' \xc1o\xdfu9?\x02C\xa5') was read. URL: https://httpbin.org/stream-bytes/10
DEBUG    api.py:603 A chunk (b' \xc1o\xdfu9?\x02C\xa5') was read. URL: https://httpbin.org/stream-bytes/10
DEBUG    api.py:603 A chunk (b' \xc1o\xdfu9?\x02C\xa5') was read. URL: https://httpbin.org/stream-bytes/10
DEBUG    api.py:603 A chunk (b' \xc1o\xdfu9?\x02C\xa5') was read. URL: https://httpbin.org/stream-bytes/10
DEBUG    api.py:603 A chunk (b' \xc1o\xdfu9?\x02C\xa5') was read. URL: https://httpbin.org/stream-bytes/10
DEBUG    api.py:603 A chunk (b' \xc1o\xdfu9?\x02C\xa5') was read. URL: https://httpbin.org/stream-bytes/10 

... it runs forever...
```
### Expected behavior
Maybe 'readany()'  must return nothing when no bytes left. It is required to take a look how **aiohttp** handle this.

### Steps to reproduce the behavior
Try to request https://file-examples.com/wp-content/uploads/2017/10/file_example_JPG_2500kB.jpg and https://httpbin.org/stream-bytes/10. Iterate over 'readany()'  and make sure that it stops when no bytes left.

### Workarounds
I do not know.

### Environment
- aiohttp-client-cache version: 0.5.0
- Python version: 3.9
- Platform: Lubuntu 20.04

**Comment by AIGeneratedUsername** [comment]
https://httpbin.org/stream-bytes/10 returns b'' when I skip the URL in the:

            urls_expire_after={
                "httpbin.org": 0, 
            },

**Comment by JWCook** [comment]
Thanks for the details. Looks like I neglected to test `readany()` (only `__aiter__()`, `iter_chunks()`, `iter_chunked()`, and `iter_any()`).

It appears that once the end of the stream is reached, `ClientResponse.content._eof` is set to `True`, and the reader returns `b''` for any calls after that. I'll work on this some more and get back to you.

**Comment by JWCook** [comment]
@AIGeneratedUsername 
Okay, that was fairly simple to fix. The problem was just that the `response.content` property was returning a new stream reader with every call. Now it will read `b''` once all the content has been read.

Give it a try again and let me know if that works for you.

**Comment by AlekseiMarinichenko** [comment]
After this update, now there are two problems:

1. `await response.content.readany()` always returns `b''` 
2. The example is broken https://aiohttp-client-cache.readthedocs.io/en/latest/examples.html#logging-requests
```
INFO:aiohttp_client_cache.backends.base:Clearing cache
Traceback (most recent call last):
  File "/backend/_local.py", line 45, in <module>
    asyncio.run(main())
  File "/usr/lib/python3.9/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/lib/python3.9/asyncio/base_events.py", line 642, in run_until_complete
    return future.result()
  File "/backend/_local.py", line 40, in main
    response = await session.get('http://httpbin.org/get')
  File "/backend/venv/lib/python3.9/site-packages/forge/_revision.py", line 322, in inner
    return await callable(*mapped.args, **mapped.kwargs)
  File "/backend/venv/lib/python3.9/site-packages/aiohttp_client_cache/session.py", line 44, in _request
    await self.cache.save_response(new_response, actions)
  File "/backend/venv/lib/python3.9/site-packages/aiohttp_client_cache/backends/base.py", line 160, in save_response
    cached_response = await CachedResponse.from_client_response(response, actions.expires)
  File "/backend/venv/lib/python3.9/site-packages/aiohttp_client_cache/response.py", line 84, in from_client_response
    client_response.content = CachedStreamReader(client_response._body)
AttributeError: can't set attribute
```

**Comment by AIGeneratedUsername** [comment]
After this update, now there are two problems:

1. `await response.content.readany()` always returns `b''` (it does not returns chunks forever as before, but it returns `b''` only once and that is all).
2. The example is broken https://aiohttp-client-cache.readthedocs.io/en/latest/examples.html#logging-requests
```
INFO:aiohttp_client_cache.backends.base:Clearing cache
Traceback (most recent call last):
  File "/backend/_local.py", line 45, in <module>
    asyncio.run(main())
  File "/usr/lib/python3.9/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/lib/python3.9/asyncio/base_events.py", line 642, in run_until_complete
    return future.result()
  File "/backend/_local.py", line 40, in main
    response = await session.get('http://httpbin.org/get')
  File "/backend/venv/lib/python3.9/site-packages/forge/_revision.py", line 322, in inner
    return await callable(*mapped.args, **mapped.kwargs)
  File "/backend/venv/lib/python3.9/site-packages/aiohttp_client_cache/session.py", line 44, in _request
    await self.cache.save_response(new_response, actions)
  File "/backend/venv/lib/python3.9/site-packages/aiohttp_client_cache/backends/base.py", line 160, in save_response
    cached_response = await CachedResponse.from_client_response(response, actions.expires)
  File "/backend/venv/lib/python3.9/site-packages/aiohttp_client_cache/response.py", line 84, in from_client_response
    client_response.content = CachedStreamReader(client_response._body)
AttributeError: can't set attribute
```

**Comment by JWCook** [comment]
`CachedResponse.content` is now settable again, which fixes #2. I can't seem to reproduce #1, though. Here's the complete test I'm running:

```python
import asyncio

from aiohttp_client_cache import CachedSession, SQLiteBackend

cache = SQLiteBackend('stream_test')


async def test_stream():
    async with CachedSession(cache=cache) as session:
        response = await session.get('https://httpbin.org/stream-bytes/10')
        print('Cached response:', response.from_cache)

        response_bytes = b""
        while chunk := await response.content.readany():
            print('  Chunk:', chunk)
            response_bytes += chunk
        print('Complete response body:', response_bytes)


async def main():
    await cache.clear()
    await test_stream()
    await test_stream()


if __name__ == '__main__':
    asyncio.run(main())
```
Which gives me the same results for both the original and the cached response.

Could you show me a code example where `await response.content.readany()` only returns `b''`? And thanks for being patient and helping me debug this.

**Comment by AIGeneratedUsername** [comment]
See https://github.com/JWCook/aiohttp-client-cache/issues/94 and https://github.com/JWCook/aiohttp-client-cache/issues/95
